File: packages/pi-session-deck/extensions/session-deck/iterm2-python-bridge.ts
import net from 'node:net';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

const DEFAULT_BRIDGE_TIMEOUT_MS = 400;
const BRIDGE_SOCKET_ENV = 'PI_SESSION_DECK_ITERM2_BRIDGE_SOCKET';

export type Iterm2PythonBridgeOpenResult =
  | { ok: true; reason: 'requested'; message: string }
  | {
      ok: false;
      reason:
        | 'python-bridge-unavailable'
        | 'automation-denied'
        | 'terminal-target-missing'
        | 'open-failed';
      message: string;
      requestSent?: boolean;
    };

export type Iterm2PythonBridgeOpenRequest =
  | { tmuxAttachArgv: readonly string[] }
  | { itermSessionId: string };

type BridgeSocket = NodeJS.ReadWriteStream & {
  destroy?: () => void;
  setEncoding?: (encoding: BufferEncoding) => void;
};

export interface Iterm2PythonBridgeClientOptions {
  env?: NodeJS.ProcessEnv;
  socketPath?: string;
  timeoutMs?: number;
  createConnection?: (path: string) => BridgeSocket;
  setTimeout?: typeof setTimeout;
  clearTimeout?: typeof clearTimeout;
}

export async function openWithIterm2PythonBridge(
  request: Iterm2PythonBridgeOpenRequest,
  options: Iterm2PythonBridgeClientOptions = {},
): Promise<Iterm2PythonBridgeOpenResult> {
  const validation = validateBridgeOpenRequest(request);
  if (!validation.ok) {
    return validation.result;
  }

  const socketPath = options.socketPath ?? getIterm2PythonBridgeSocketPath(options.env);
  const timeoutMs = options.timeoutMs ?? DEFAULT_BRIDGE_TIMEOUT_MS;
  const createConnection =
    options.createConnection ?? ((path: string): BridgeSocket => net.createConnection(path));
  const setTimer = options.setTimeout ?? setTimeout;
  const clearTimer = options.clearTimeout ?? clearTimeout;

  return new Promise<Iterm2PythonBridgeOpenResult>((resolve) => {
    let socket: BridgeSocket;
    try {
      socket = createConnection(socketPath);
    } catch (error) {
      resolve({
        ok: false,
        reason: 'python-bridge-unavailable',
        message: `iTerm2 Python bridge is unavailable: ${getErrorMessage(error)}`,
        requestSent: false,
      });
      return;
    }

    let settled = false;
    let requestSent = false;
    let buffer = '';

    const finish = (result: Iterm2PythonBridgeOpenResult) => {
      if (settled) {
        return;
      }
      settled = true;
      clearTimer(timeout);
      try {
        socket.end();
      } catch {
        // Best effort cleanup.
      }
      resolve(result);
    };

    const timeout = setTimer(() => {
      try {
        socket.destroy?.();
      } catch {
        // Best effort cleanup.
      }
      finish({
        ok: false,
        reason: 'python-bridge-unavailable',
        message: 'iTerm2 Python bridge did not respond before the timeout.',
        requestSent,
      });
    }, timeoutMs);

    socket.setEncoding?.('utf8');
    socket.on('connect', () => {
      try {
        socket.write(`${JSON.stringify(request)}\n`);
        requestSent = true;
      } catch (error) {
        finish({
          ok: false,
          reason: 'python-bridge-unavailable',
          message: `iTerm2 Python bridge could not send the request: ${getErrorMessage(error)}`,
          requestSent: false,
        });
      }
    });
    socket.on('data', (chunk) => {
      buffer += String(chunk);
      const newlineIndex = buffer.indexOf('\n');
      if (newlineIndex === -1) {
        return;
      }
      const line = buffer.slice(0, newlineIndex);
      finish(parseBridgeResponse(line, requestSent, validation.successMessage));
    });
    socket.on('error', (error) => {
      finish({
        ok: false,
        reason: 'python-bridge-unavailable',
        message: `iTerm2 Python bridge is unavailable: ${getErrorMessage(error)}`,
        requestSent,
      });
    });
    socket.on('close', () => {
      finish({
        ok: false,
        reason: 'python-bridge-unavailable',
        message: 'iTerm2 Python bridge closed before returning a response.',
        requestSent,
      });
    });
  });
}

export function getIterm2PythonBridgeSocketPath(env: NodeJS.ProcessEnv = process.env): string {
  const configured = trimNonEmpty(env[BRIDGE_SOCKET_ENV]);
  if (configured !== undefined) {
    return configured;
  }

  const uid = typeof process.getuid === 'function' ? process.getuid() : 'user';
  return join(tmpdir(), `pi-session-deck-${uid}`, 'iterm2-python-bridge.sock');
}

function parseBridgeResponse(
  line: string,
  requestSent: boolean,
  successMessage: string,
): Iterm2PythonBridgeOpenResult {
  let parsed: unknown;
  try {
    parsed = JSON.parse(line) as unknown;
  } catch (error) {
    return {
      ok: false,
      reason: 'open-failed',
      message: `iTerm2 Python bridge returned malformed JSON: ${getErrorMessage(error)}`,
      requestSent,
    };
  }

  if (!isObject(parsed)) {
    return {
      ok: false,
      reason: 'open-failed',
      message: 'iTerm2 Python bridge returned a malformed response.',
      requestSent,
    };
  }

  if (parsed['ok'] === true) {
    return {
      ok: true,
      reason: 'requested',
      message: typeof parsed['message'] === 'string' ? parsed['message'] : successMessage,
    };
  }

  const reason = normalizeBridgeFailureReason(parsed['reason']);
  return {
    ok: false,
    reason,
    message:
      typeof parsed['message'] === 'string' && parsed['message'].length > 0
        ? parsed['message']
        : 'iTerm2 Python bridge failed to open the tmux attach tab.',
    requestSent,
  };
}

function normalizeBridgeFailureReason(
  reason: unknown,
): Extract<Iterm2PythonBridgeOpenResult, { ok: false }>['reason'] {
  switch (reason) {
    case 'python-bridge-unavailable':
    case 'automation-denied':
    case 'terminal-target-missing':
    case 'open-failed':
      return reason;
    default:
      return 'open-failed';
  }
}

function validateBridgeOpenRequest(
  request: Iterm2PythonBridgeOpenRequest,
): { ok: true; successMessage: string } | { ok: false; result: Iterm2PythonBridgeOpenResult } {
  if ('tmuxAttachArgv' in request) {
    if (isValidTmuxAttachArgv(request.tmuxAttachArgv)) {
      return { ok: true, successMessage: 'Requested tmux attach in a new iTerm2 tab.' };
    }

    return {
      ok: false,
      result: {
        ok: false,
        reason: 'open-failed',
        message: 'Refusing to send an invalid tmux attach argv to the iTerm2 bridge.',
        requestSent: false,
      },
    };
  }

  if (isNonBlankString(request.itermSessionId)) {
    return { ok: true, successMessage: 'Requested iTerm2 focus for selected session.' };
  }

  return {
    ok: false,
    result: {
      ok: false,
      reason: 'open-failed',
      message: 'Refusing to send an invalid iTerm2 session id to the iTerm2 bridge.',
      requestSent: false,
    },
  };
}

function isValidTmuxAttachArgv(argv: unknown): argv is readonly string[] {
  if (!Array.isArray(argv) || argv.length !== 7 || !argv.every(isNonBlankString)) {
    return false;
  }

  const [command, selectorFlag, selectorValue, attachSession, keepEnvironment, targetFlag] =
    argv as [string, string, string, string, string, string, string];
  if (command !== 'tmux') {
    return false;
  }

  if (selectorFlag === '-L' && selectorValue.includes('/')) {
    return false;
  }

  if (selectorFlag !== '-S' && selectorFlag !== '-L') {
    return false;
  }

  return attachSession === 'attach-session' && keepEnvironment === '-E' && targetFlag === '-t';
}

function isNonBlankString(value: unknown): value is string {
  return typeof value === 'string' && value.trim().length > 0;
}

function trimNonEmpty(value: unknown): string | undefined {
  if (typeof value !== 'string') {
    return undefined;
  }

  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

function isObject(candidate: unknown): candidate is Record<string, unknown> {
  return typeof candidate === 'object' && candidate !== null;
}

function getErrorMessage(error: unknown): string {
  if (error instanceof Error && error.message.length > 0) {
    return error.message;
  }

  return String(error);
}

File: packages/pi-session-deck/extensions/session-deck/iterm2-python-bridge.py
#!/usr/bin/env python3
"""iTerm2 AutoLaunch bridge for pi-session-deck terminal focus.

Install by symlinking or copying this file into iTerm2's Scripts/AutoLaunch
folder. The TypeScript side talks to a user-local Unix socket using JSON-lines:

  {"tmuxAttachArgv":["tmux","-S","...","attach-session","-E","-t","..."]}\n
or:

  {"itermSessionId":"w0t0p0:..."}\n
Only exact tmux attach argv arrays and non-empty iTerm2 session ids are accepted.
"""

from __future__ import annotations

import asyncio
import json
import os
import shlex
import signal
import tempfile
from pathlib import Path
from typing import Any, Optional

import iterm2

SOCKET_ENV = "PI_SESSION_DECK_ITERM2_BRIDGE_SOCKET"
REQUEST_TIMEOUT_SECONDS = 2.0
INVALID_ATTACH_ARGV_MESSAGE = (
    "Bridge only accepts exact tmux attach argv in 'tmuxAttachArgv'."
)
INVALID_ITERM_SESSION_ID_MESSAGE = (
    "Bridge only accepts non-empty iTerm2 session ids in 'itermSessionId'."
)


def default_socket_path() -> Path:
    configured = os.environ.get(SOCKET_ENV)
    if configured:
        return Path(configured).expanduser()

    uid = os.getuid() if hasattr(os, "getuid") else "user"
    return Path(tempfile.gettempdir()) / f"pi-session-deck-{uid}" / "iterm2-python-bridge.sock"


def prepare_socket_path(path: Path) -> None:
    path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
    try:
        path.parent.chmod(0o700)
    except OSError:
        pass

    try:
        path.unlink()
    except FileNotFoundError:
        pass


def response(ok: bool, **fields: Any) -> bytes:
    return (json.dumps({"ok": ok, **fields}, separators=(",", ":")) + "\n").encode("utf-8")


def validate_tmux_attach_argv(candidate: Any) -> Optional[list[str]]:
    if not isinstance(candidate, list) or len(candidate) != 7:
        return None
    if any(not isinstance(arg, str) or len(arg.strip()) == 0 for arg in candidate):
        return None

    program, selector_flag, selector, subcommand, preserve_env, target_flag, _target = candidate
    if (
        program != "tmux"
        or selector_flag not in ("-S", "-L")
        or subcommand != "attach-session"
        or preserve_env != "-E"
        or target_flag != "-t"
    ):
        return None
    if selector_flag == "-L" and "/" in selector:
        return None

    return list(candidate)


def format_iterm2_shell_command(tmux_attach_argv: list[str]) -> str:
    return "exec " + " ".join(shlex.quote(arg) for arg in tmux_attach_argv)


def validate_iterm_session_id(candidate: Any) -> Optional[str]:
    if not isinstance(candidate, str):
        return None

    session_id = candidate.strip()
    return session_id if session_id else None


async def maybe_call(action: Any) -> None:
    if action is None:
        return
    await action()


async def open_iterm2_tab(connection: iterm2.Connection, tmux_attach_argv: list[str]) -> None:
    command = format_iterm2_shell_command(tmux_attach_argv)
    app = await iterm2.async_get_app(connection)
    window = app.current_terminal_window

    if window is None:
        window = await iterm2.Window.async_create(connection, command=command)
        tab = window.current_tab
    else:
        tab = await window.async_create_tab(command=command)

    for action in (
        getattr(app, "async_activate", None),
        getattr(window, "async_activate", None),
        getattr(tab, "async_select", None),
    ):
        if action is None:
            continue
        try:
            await action()
        except Exception:
            pass


async def focus_iterm2_session(connection: iterm2.Connection, iterm_session_id: str) -> bool:
    app = await iterm2.async_get_app(connection)
    get_session_by_id = getattr(app, "get_session_by_id", None)
    if get_session_by_id is None:
        raise RuntimeError("iTerm2 Python API does not expose get_session_by_id")

    candidate_ids = [iterm_session_id]
    if ":" in iterm_session_id:
        candidate_ids.append(iterm_session_id.rsplit(":", 1)[1])

    session = None
    for candidate_id in candidate_ids:
        try:
            session = get_session_by_id(candidate_id)
        except Exception:
            session = None
        if session is not None:
            break

    if session is None:
        return False

    tab = getattr(session, "tab", None)
    window = getattr(tab, "window", None) if tab is not None else None

    for action in (
        getattr(app, "async_activate", None),
        getattr(window, "async_activate", None),
        getattr(tab, "async_select", None),
    ):
        try:
            await maybe_call(action)
        except Exception:
            pass

    activate_session = getattr(session, "async_activate", None)
    if activate_session is None:
        raise RuntimeError("iTerm2 Python API does not expose Session.async_activate")
    await activate_session()
    return True


async def handle_client(
    connection: iterm2.Connection,
    reader: asyncio.StreamReader,
    writer: asyncio.StreamWriter,
) -> None:
    try:
        line = await asyncio.wait_for(reader.readline(), timeout=REQUEST_TIMEOUT_SECONDS)
        request = json.loads(line.decode("utf-8"))
        if not isinstance(request, dict):
            writer.write(
                response(False, reason="open-failed", message="Bridge request must be an object.")
            )
            await writer.drain()
            return

        tmux_attach_argv = validate_tmux_attach_argv(request.get("tmuxAttachArgv"))
        iterm_session_id = validate_iterm_session_id(request.get("itermSessionId"))

        if tmux_attach_argv is not None:
            await open_iterm2_tab(connection, tmux_attach_argv)
            writer.write(response(True, message="Requested tmux attach in a new iTerm2 tab."))
        elif iterm_session_id is not None:
            focused = await focus_iterm2_session(connection, iterm_session_id)
            if focused:
                writer.write(response(True, message="Requested iTerm2 focus for selected session."))
            else:
                writer.write(
                    response(
                        False,
                        reason="terminal-target-missing",
                        message="iTerm2 session is no longer available.",
                    )
                )
        else:
            message = (
                INVALID_ATTACH_ARGV_MESSAGE
                if "tmuxAttachArgv" in request
                else INVALID_ITERM_SESSION_ID_MESSAGE
            )
            writer.write(response(False, reason="open-failed", message=message))
        await writer.drain()
    except asyncio.TimeoutError:
        writer.write(
            response(False, reason="open-failed", message="Bridge request timed out.")
        )
        await writer.drain()
    except json.JSONDecodeError as error:
        writer.write(
            response(False, reason="open-failed", message=f"Malformed JSON request: {error}")
        )
        await writer.drain()
    except Exception as error:  # iTerm2 API failures should be reported, not crash the daemon.
        writer.write(
            response(False, reason="open-failed", message=f"Failed to request iTerm2 focus: {error}")
        )
        await writer.drain()
    finally:
        writer.close()
        try:
            await writer.wait_closed()
        except Exception:
            pass


async def main(connection: iterm2.Connection) -> None:
    socket_path = default_socket_path()
    prepare_socket_path(socket_path)

    server = None

    def cleanup() -> None:
        try:
            socket_path.unlink()
        except FileNotFoundError:
            pass

    try:
        server = await asyncio.start_unix_server(
            lambda reader, writer: handle_client(connection, reader, writer),
            path=str(socket_path),
        )

        loop = asyncio.get_running_loop()
        shutdown_requested = asyncio.Event()

        def request_shutdown() -> None:
            loop.call_soon_threadsafe(shutdown_requested.set)

        for sig in (signal.SIGINT, signal.SIGTERM):
            try:
                loop.add_signal_handler(sig, request_shutdown)
            except (NotImplementedError, RuntimeError):
                pass

        await shutdown_requested.wait()
    finally:
        if server is not None:
            server.close()
            await server.wait_closed()
        cleanup()


iterm2.run_forever(main)
